### PR TITLE
execute SQL on python objects

### PR DESCRIPTION
**Pull request by barakalon**
My brain is fried right now, so I should do another pass on cleaning this up tomorrow.

But I wanted to put up my progress so far.

### CODE REVIEW

File: sqlglot/executor/python.py [hunk 1]
```diff
@@ -164,24 +181,38 @@ def join(self, step, context):
         condition = self.generate(step.condition)
         projections = self.generate_tuple(step.projections)
 
-        if not condition or not projections:
+        if not condition and not projections:
             return source_context
+        elif not projections:
```
Comment (barakalon) [inline sqlglot/executor/python.py:186]: This all feels pretty messy to me.

Without projections, we need to keep the nested tables in the resulting context.

File: sqlglot/executor/__init__.py [hunk 2]
```diff
@@ -23,9 +24,11 @@ def execute(sql, schema, read=None):
                 3. {catalog: {db: {table: {col: type}}}}
         read (str): the SQL dialect to apply during parsing
             (eg. "spark", "hive", "presto", "mysql").
+        tables (dict[str, list[dict]|sqlglot.executor.Table]): additional tables to register
     Returns:
         sqlglot.executor.Table: Simple columnar data structure.
     """
+    tables = _ensure_tables(tables)
```
Comment (tobymao) [inline sqlglot/executor/__init__.py:31]: given tables, you should be able to infer the schema

File: sqlglot/executor/__init__.py [hunk 3]
```diff
@@ -34,6 +37,20 @@ def execute(sql, schema, read=None):
...
+
+def _ensure_tables(tables):
+    result = {}
+    if not tables:
+        return result
+    for name, table in tables.items():
+        if isinstance(table, Table):
+            result[name] = table
+        else:
+            columns = tuple(table[0].keys()) if table else ()
```
Comment (tobymao) [inline sqlglot/executor/__init__.py:53]: you don't need keys

File: sqlglot/executor/python.py (hunk 1 above)
Comment (tobymao) [reply sqlglot/executor/python.py:186]: i don't get this, you may have to explain it to me

File: sqlglot/executor/__init__.py (hunk 2 above)
Comment (barakalon) [reply sqlglot/executor/__init__.py:31]: Unless the tables are empty

File: sqlglot/executor/__init__.py (hunk 3 above)
Comment (barakalon) [reply sqlglot/executor/__init__.py:53]: 🤦 how many times have you told me this? lol

File: sqlglot/executor/python.py (hunk 1 above)
Comment (barakalon) [reply sqlglot/executor/python.py:186]: let's chat today

**Review by barakalon — COMMENTED** [review]
@tobymao Alright I got this working

**Pull request merged**